### DISCUSSION TITLE
Heated channel with fluid not exiting at outlet

### DISCUSSION THREAD

**Discussion by oh304** [idaholab/moose#27761] [Q&A Modules: Navier-Stokes] [answered]
### Check these boxes if you have followed the posting rules.

- [X] Q&A Navier Stokes is the most appropriate category for my question
- [X] I have consulted the posting Guidelines on the Discussions front page
- [X] I have searched the Discussions forum and my question has not been asked before
- [X] I have searched the MOOSE website and the documentation does not answer my question
- [X] I have formatted my post following the posting guidelines (avoid screenshots if possible, triple back quotes before/after pasted text, etc)

### Question

Hi there,
I have been using the Navier Stokes module to simulate a horizontal coolant channel (with inconel walls) being heated up using a FVDirichletBC boundary condition on the inconel. Heat is transferred through a FVConvectionCorrelationInterface to the coolant, which travels along the x direction. The inlet coolant boundary has a temperature defined by a heat flux, and inlet velocity defined with the INSFVInletVelocityBC.

My issue is, at the coolant outlet, the velocity of the fluid speeds up at the bottom, and appears to turn to move back on itself. It's as if the coolant if 'hitting' the outlet boundary, and turning back on itself. This can be seen in the image below. Am I missing something very obvious? Thanks :)

![Screenshot 2024-05-31 144644](https://github.com/idaholab/moose/assets/134851749/1c3add22-da97-4605-9617-4c679581cb79)

I have set up the following boundary conditions:
```
[FVBCs]
    [inlet-u]
      type = INSFVInletVelocityBC
      boundary = 'inlet'
      variable = u
      function = ${inlet_velocity}
    []
    [inlet-v]
      type = INSFVInletVelocityBC
      boundary = 'inlet'
      variable = v
      function = 0
    []
    [inlet-T]
      type = FVNeumannBC
      variable = T
      value = '${fparse rho * cp * inlet_velocity * inlet_temp}'
      boundary = 'inlet'
    []
    [no-slip-u]
      type = INSFVNoSlipWallBC
      boundary = 'inner_wall_top inner_wall_bottom'
      variable = u
      function = 0.000001
    []
    [no-slip-v]
      type = INSFVNoSlipWallBC
      boundary = 'inner_wall_top inner_wall_bottom'
      variable = v
      function = 0.000001
    []
    [bottom]
        type = FVDirichletBC
        variable = Ts
        boundary = outer_wall_bottom
        value = 700
    []
    [top]
        type = FVDirichletBC
        variable = Ts
        boundary = outer_wall_top
        value = 700
    []
    [outlet_momentum_u]
      type = INSFVMomentumAdvectionOutflowBC
      variable = u
      u = u
      v = v
      boundary = 'outlet'
      momentum_component = 'x'
      rho = ${rho}
    []
    [outlet_momentum_v]
      type = INSFVMomentumAdvectionOutflowBC
      variable = v
      u = u
      v = v
      boundary = 'outlet'
      momentum_component = 'y'
      rho = ${rho}
    []
    [outlet_p]
      type = INSFVOutletPressureBC
      boundary = 'outlet'
      variable = pressure
      function = '${outlet_pressure}'
    []
[]
```
### Additional information

Mesh size and type:
 Reynolds number:
 Discretization (finite element CG/DG, finite volume, etc):
 Models (turbulence, porous media, etc):
 Solver method (fully coupled, segregated, multiapps, etc):
 Base input you started from:

**Comment by GiudGiud** [comment]
Hello

This does look weird indeed.
Is this the steady state solution? Such a state could be an intermediate state moving from the inital condition to the solution we expect

**Reply by oh304** [reply]
Thank you for the reply! The simulation is transient, and the effect is there at every time step. Could I be missing something else?

Best wishes,



Attached is my executioner:
```
[Executioner]
    type = Transient
    end_time = 300
    dtmin = 1e-8
    dtmax = 0.1
    petsc_options = '-snes_converged_reason -ksp_converged_reason -options_left'
    solve_type = 'NEWTON'
    line_search = 'none'
    nl_max_its = 15
    l_max_its = 500
    nl_rel_tol = 1e-10
    [TimeStepper]
      type = IterationAdaptiveDT
      optimal_iterations = 7
      dt = 0.5
      linear_iteration_ratio = 1e6
      growth_factor = 1.5
    []
[]
```

**Reply by GiudGiud** [reply]
I m not sure what we could be missing. 
Can you share your input and your mesh on a github repository?

**Reply by oh304** [reply]
That'd be great! Thanks for having a look.

https://github.com/oh304/moose_sandbox

**Reply by GiudGiud** [reply]
@lindsayad or @freiler if you have time to take a look

**Reply by lindsayad** [reply]
Could you create your mesh with quad elements instead of tris? FV appreciates as little skew as possible

**Reply by lindsayad** [reply]
The artifact does seem to go away if you remove the Boussinesq kernels for the momentum equation. Do you need those kernels? Do you expect natural convection to be significant compared to the forced convection you have in your channel flow?

**Reply by tanoret** [reply]
There is a problem with the BCs, the Boussinesq term gives you a pressure gradient in the vertical direction, but the outlet BC specifies a constant pressure at the outlet, which is not consistent with the vertical pressure gradient due to buoyancy. This results in an acceleration at the outlet.

Some solutions might be:
- Specify a pressure field at the outlet that matches the expected pressure gradient due to buoyancy.
- Create a boundary condition that corrects the pressure at the outlet for thermal buoyancy. 

Note: it does not mean that the flow is not exiting...

**Reply by tanoret** [reply, marked as answer]
This is what you need to do for the solution in the first bullet point

```
...
[AuxVariables]
    ...
    [hydrostatic_pressure]
      type = MooseVariableFVReal
    []
    ...
[]
...
...
[AuxKernels]
    ...
    [populate_hydro_pressure]
      type = ParsedAux
      variable = hydrostatic_pressure
      coupled_variables = T
      use_xyzt = true
      expression = '-${alpha} * ${g} * ${rho} * y * (T - ${reference_temperature}) + ${outlet_pressure}'
      block = 'coolant'
    []
    ...
[]
...
...
[FVBCs]
    ...
    [outlet_p]
      type = INSFVOutletPressureBC
      boundary = 'outlet'
      variable = pressure
      functor = 'hydrostatic_pressure'
    []
    ...
[]
...
```

This is the pressure profile with the previous solution (which introduces the artificial acceleration at the outlet):

<img width="732" alt="image" src="https://github.com/idaholab/moose/assets/29891082/fecd9507-7a10-42e7-811c-e7ce172cac0b">

.. and this is the one with the new one:

<img width="732" alt="image" src="https://github.com/idaholab/moose/assets/29891082/38f5900e-5988-4f56-8fc4-468072d0b45d">

Please also consider quad elements as @lindsayad suggested, since the velocity profile is not really smooth in the tri mesh:

<img width="732" alt="image" src="https://github.com/idaholab/moose/assets/29891082/56f7d23d-a5da-484b-a3ae-1020bb146693">

**Reply by oh304** [reply]
Ah! Thank you so much for your responses @tanoret @lindsayad @GiudGiud - much appreciated! That pressure outlet definitely looks better with the hydro-static pressure functor! I will implement these changes now, including the quad elements :)

Thanks again!
Oscar

**Reply by lindsayad** [reply]
Since you are doing a monolithic solve with newton, I would suggest changing the aux  variable/aux kernel to a functor material, so
```
[FunctorMaterials]
    ...
    [populate_hydro_pressure]
      type = ADParsedFunctorMaterial
      property_name = hydrostatic_pressure
      functor_names = 'T'
      use_xyzt = true
      expression = '-${alpha} * ${g} * ${rho} * y * (T - ${reference_temperature}) + ${outlet_pressure}'
      block = 'coolant'
    []
    ...
 ```
since the aux kernel will drop derivatives from `T`
